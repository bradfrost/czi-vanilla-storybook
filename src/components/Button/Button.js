import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import Icon from '../Icon';
import './Button.scss';
import '../Utilities/Utilities.scss';

function Button(props) {
  const {
    className,
    variant,
    size,
    theme,
    disabled,
    fullWidth,
    iconName,
    iconPosition,
    loading,
    screenReaderText,
    buttonRef,
    href,
    text,
    type,
    hideText,
    ...other
  } = props;

  const componentClassName = classnames('cn-c-btn', className, {
    'cn-c-btn--primary': variant == 'primary',
    'cn-c-btn--tertiary': variant == 'tertiary',
    'cn-c-btn--ghost': variant == 'ghost',
    'cn-c-btn--bare': variant == 'bare',
    'cn-c-btn--link': variant == 'link',
    'cn-c-btn--large': size == 'large',
    'cn-c-btn--small': size == 'small',
    'cn-c-btn--tiny': size == 'tiny',
    'cn-c-btn--inverted': theme == 'inverted',
    'cn-c-btn--block': fullWidth,
    'cn-is-loading': loading
  });

  const TagName = href ? 'a' : 'button';
  return (
    <TagName
      className={componentClassName}
      href={href}
      disabled={disabled}
      tabIndex={disabled ? -1 : undefined}
      ref={buttonRef}
      type={type}
      {...other}
    >
      {loading && (
        <Icon
          aria-hidden="true"
          focusable="false"
          name="test-loader"
          className="cn-c-btn__icon"
        />
      )}
      {iconPosition === 'before' && (
        <Icon
          aria-hidden="true"
          focusable="false"
          name={iconName}
          className="cn-c-btn__icon"
        />
      )}
      {text && (
        <span
          className={
            !hideText ? 'cn-c-btn__text' : 'cn-c-btn__text cn-u-is-vishidden'
          }
        >
          {text}
          {screenReaderText && (
            <span className="cn-u-is-vishidden">{screenReaderText}</span>
          )}
        </span>
      )}
      {iconPosition === 'after' && (
        <Icon
          aria-hidden="true"
          focusable="false"
          name={iconName}
          className="cn-c-btn__icon"
        />
      )}
    </TagName>
  );
}

Button.propTypes = {
  /**
   * Reference to the button DOM node
   */
  buttonRef: PropTypes.object,
  /**
   * CSS class names that can be appended to the component.
   */
  className: PropTypes.string,
  /**
   * Disables the field and prevents editing the contents
   */
  disabled: PropTypes.bool,
  /**
   * Toggles button that fills the full width of its container
   */
  fullWidth: PropTypes.bool,
  /**
   * Visually hide button text (but text is still accessible to assistive technology)
   */
  hideText: PropTypes.bool,
  /**
   * Link to URL. If href is present, the button will be rendered as an <a> element.
   */
  href: PropTypes.PropTypes.string,
  /**
   * Name of SVG icon (i.e. caret-down, minus, warning)
   */
  iconName: PropTypes.string,
  /**
   * Determines position of icon relative to button text.
   * - **before** places icon before button text
   * - **after** places icon after button text
   */
  iconPosition: PropTypes.oneOf(['before', 'after']),
  /**
   * Loading state passed down from higher level used to trigger loader and text change
   */
  loading: PropTypes.bool,
  /**
   * Visually hidden additional instruction text to help provide screen reader users additional context. For instance, "View details" might be the visible button text, but screenReaderText might add additional instructions such as "for confirmation number C1234567"
   */
  screenReaderText: PropTypes.string,
  /**
   * Available size variations for the button
   */
  size: PropTypes.oneOf(['tiny', 'small', 'large']),
  /**
   * The visible button text
   */
  text: PropTypes.string.isRequired,
  /**
   * Available theme variations for the button
   */
  theme: PropTypes.oneOf(['inverted']),
  /**
   * Determines type of button
   * - **button** The button is a clickable button.
   * - **submit** The button is a submit button (submits form data). This is the default `button` behavior
   * - **reset** The button is a reset button (resets the form-data to its initial values)
   */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  /**
   * Available _stylistic_ variations available for the Button component
   */
  variant: PropTypes.oneOf([
    'primary',
    'secondary',
    'tertiary',
    'bare',
    'link',
    'ghost'
  ])
};

Button.defaultProps = {
  disabled: false,
  fullWidth: false,
  hideText: false
};

export default Button;
