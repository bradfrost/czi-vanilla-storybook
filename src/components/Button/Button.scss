@import '../../css/scss/abstracts/variables';
@import '../../css/scss/abstracts/mixins';

/*------------------------------------*\
    #BUTTON
\*------------------------------------*/

/**
 * 1) Button or link that has functionality to it
 * 2) Defaults to primary button
 * 3) Margin 0 added becuase Safari has margin on the button
 */
.cn-c-btn {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding-top: size(1.5);
  padding-right: size(5);
  padding-bottom: size(1.5);
  padding-left: size(5);
  margin: 0;
  border-width: $border-width-med;
  border-style: solid;
  border-color: $color-brand-blue;
  border-radius: $border-radius;
  @include typography-preset-5--bold;
  font-family: inherit;
  background-color: $color-neutral-white;
  color: $color-brand-blue;
  cursor: pointer;
  transition: background-color $anim-fade-quick $anim-ease,
    border-color $anim-fade-quick $anim-ease,
    box-shadow $anim-fade-quick $anim-ease,
    transform $anim-fade-quick $anim-ease;

  &:hover,
  &:focus {
    background-color: $color-neutral-white;
    color: $color-tint-brand-atlantic-amethyst-shade;
    border-color: $color-tint-brand-atlantic-amethyst-shade;
  }

  &:focus {
    @include focus;
  }

  &:active {
    background: $color-neutral-gray-05;
  }

  /**
   * Button within vertical button group
   */
  .cn-c-btn-group--vertical & {
    display: block;
    margin-bottom: size(1);
  }

  /**
   * Disabled buttons
   */
  &:disabled {
    background: $color-neutral-gray-20;
    border-color: $color-neutral-gray-20;
    color: $color-neutral-white;
    text-shadow: none;
    cursor: not-allowed;

    &:hover,
    &:focus {
      background: $color-neutral-gray-20;
      border-color: $color-neutral-gray-20;
      box-shadow: none;
      transform: none;
    }
  }

  /**
   * Loading button
   */
  &.cn-is-loading {
    opacity: 0.65;
    cursor: not-allowed;

    &:hover,
    &:focus {
      color: $color-brand-blue;
      border-color: $color-brand-blue;
    }
  }
}

/**
 * Primary button
 * 1) Prominent button styles that should be applied to
 *    the most important action of a page.
 */
.cn-c-btn--primary {
  border-width: $border-width-med;
  border-style: solid;
  border-color: $color-tint-brand-atlantic-amethyst-shade-light;
  background-color: $color-primary-action-background;
  color: $color-primary-action-text;
  box-shadow: none;

  &:focus {
    @include focus;
  }

  &:active {
    transition: none;
    background-color: $color-primary-action-background-active;
  }

  /**
   * Loading primary button
   * 1) Keep the button the same colors and in the same position on hover and focus
   */
  &.cn-is-loading {
    &:hover,
    &:focus {
      box-shadow: none;
      color: $color-neutral-white;
      background-color: $color-brand-blue;
      border-color: $color-tint-brand-atlantic-amethyst-shade-light;
      transform: translate(0px, 0px);
    }
  }
}

/**
 * Small button
 * 1) Button size is smaller than the default
 * 2) Align button height with form field height using line length
 */
.cn-c-btn--small {
  @include typography-preset-7--bold;
  line-height: 1.9; /* 2 */
  padding-top: size(1);
  padding-right: size(2);
  padding-bottom: size(1);
  padding-left: size(2);
}

/**
 * Tiny button
 * 1) Button size is smaller than the default
 */
.cn-c-btn--tiny {
  padding-top: size(0.5);
  padding-right: size(1);
  padding-bottom: size(0.5);
  padding-left: size(1);
}

/**
 * Link Button
 * 1) A button that is visually styled like a text link
 */
.cn-c-btn--link {
  align-items: flex-start;
  padding: 0;
  border: 0;
  border-radius: 0;
  text-align: left;
  @include text-link();
  background-color: transparent;

  &:hover,
  &:focus {
    background-color: transparent;
  }
}

/**
 * Link inverted button
 * 1) A button that is visually styled like a text link with white text
 * 2) Used for link button on a dark background
 */
.cn-c-btn--link-inverted {
  align-items: flex-start;
  padding: 0;
  border: 0;
  border-radius: 0;
  text-align: left;
  @include text-link();
  color: $color-neutral-white;
  background-color: transparent;
  transition: color $anim-fade-quick $anim-ease;

  &:hover,
  &:focus {
    background-color: transparent;
  }
}

/**
 * Bare Button
 * 1) A button that strips out borders, backgrounds, etc
 * 2) Use for Close "X" buttons, tooltip icons, and other
 */
.cn-c-btn--bare {
  background-color: transparent;
  border: 0;
  padding: 0;

  /**
     * Bare button within inverted feature
     * 1) Change text color to white when bare button is within inverted feature
     */
  .cn-c-feature--inverted & {
    color: $color-neutral-white; /* 1 */
  }

  &:hover,
  &:focus {
    color: $color-brand-blue;
    background-color: transparent;
  }
}

/**
 * Block button
 * 1) A block button fills 100% of the width of its container
 */
.cn-c-btn--block {
  width: 100%;
}

/**
 * Ghost button
 * 1) A button that has a light border and text color used for dark backgrounds
 */
.cn-c-btn--ghost {
  background-color: transparent;
  color: $color-neutral-white;
  border: $border-width-med solid $color-neutral-white;

  &:hover,
  &:focus {
    color: $color-neutral-black;
    background-color: $color-neutral-white;
    border: $border-width-med solid $color-neutral-white;
  }
}

/**
 * Button text
 * 1) Span of text used within button used to help position on display flex
 */
.cn-c-btn__text {
  /**
     * Button text within menu button and signup button
     * 1) Visibly hide text on small screens
     * 2) Show text on medium and large screens
     */
  .cn-c-header__menu-btn &,
  .cn-c-header__signin-btn & {
    position: absolute;
    overflow: hidden;
    width: 1px;
    height: 1px;
    padding: 0;
    border: 0;
    clip: rect(1px, 1px, 1px, 1px);

    @media all and (min-width: $bp-small) {
      position: static;
      visibility: visible;
      height: auto;
      width: auto;
      clip: none;
    }
  }
}

/**
 * Button Icon
 * 1) Set SVG icon fill to the current text color
 */
.cn-c-btn__icon {
  fill: currentColor; /* 1 */

  /**
   * Button icon within link button
   * 1) Button icon should align with top line of text if link button text wraps
   */
  .cn-c-btn--link & {
    position: relative; /* 1 */
    top: 4px; /* 1 */
  }

  /**
   * Button icon within small link button
   */
  .cn-c-btn--link.cn-c-btn--small & {
    top: 2px;
  }

  /**
   * Button icon within counter button
   */
  .cn-c-counter__btn:disabled & {
    fill: $color-neutral-white;
  }

  /**
   * Button icon within header menu and signup buttons
   */
  .cn-c-header__menu-btn &,
  .cn-c-header__signin-btn & {
    height: $font-size-med-2;
    width: $font-size-med-2;

    @media all and (min-width: $bp-small) {
      height: $font-size-med;
      width: $font-size-med;
    }
  }

  /**
   * Button icon within show hide component
   * 1) Add transition for icon flip when activated
   */
  .cn-c-show-hide__btn & {
    position: relative;
    top: 8px;
    height: 10px;
    width: 10px;
    transition: transform $anim-ease $anim-fade-quick; /* 1 */
  }

  /**
   * Button icon within small show hide component
   */
  .cn-c-show-hide--small .cn-c-show-hide__btn & {
    top: 5px;
  }

  /**
   * Button icon within tiny button
   */
  .cn-c-btn--tiny & {
    height: 10px;
    width: 10px;
  }

  /**
   * Button icon within active show hide component
   * 1) Rotate icon 180 deg when show hide component is open
   */
  .cn-c-show-hide.cn-is-active .cn-c-show-hide__btn & {
    transform: rotate(90deg); /* 1 */
  }

  .cn-c-btn.cn-is-loading & {
    animation: rotateIcon 2s linear infinite;
  }
}

/**
 * Button text before icon spacing
 * 1) Controls the spacing between the button text
 *    and the icon that comes _before_ button text
 */
.cn-c-btn__icon + .cn-c-btn__text:not(.cn-u-is-vishidden) {
  margin-left: size(1); /* 1 */

  .cn-c-btn--small & {
    margin-left: size(0.5);
  }
}

/**
 * Button after icon spacing
 * 1) Controls the spacing between the button text
 *    and the icon that comes _after_ button text
 */
.cn-c-btn__text:not(.cn-u-is-vishidden) + .cn-c-btn__icon {
  margin-left: size(1); /* 1 */

  /**
   * Button icon after button text that's not vishidden within a link variant button
   */
  .cn-c-btn--link & {
    margin-left: size(0.5);
  }

  .cn-c-btn--small & {
    margin-left: size(0.5);
  }
}

/* The animation code */
@keyframes rotateIcon {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
