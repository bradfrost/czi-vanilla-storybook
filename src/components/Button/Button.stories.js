import React from 'react';
import Button from './Button';

export default {
  component: Button,
  title: 'Molecules/Buttons/Button'
};

const Template = args => <Button {...args} />;

export const Default = Template.bind({});
Default.args = { text: 'Button' };

export const DefaultWithIconBefore = Template.bind({});
DefaultWithIconBefore.args = {
  text: 'Button',
  iconPosition: 'before',
  iconName: 'caret-left'
};

export const DefaultWithIconAfter = Template.bind({});
DefaultWithIconAfter.args = {
  text: 'Button',
  iconPosition: 'after',
  iconName: 'caret-right'
};

export const Primary = Template.bind({});
Primary.args = { variant: 'primary', text: 'Primary Button' };

export const Tertiary = Template.bind({});
Tertiary.args = { variant: 'tertiary', text: 'Tertiary Button' };

export const Ghost = () => (
  <div className="cn-u-padding-large" style={{ background: '#000000' }}>
    <Button
      variant="ghost"
      text="Ghost button"
      screenReaderText="for screenreaders"
    />
  </div>
);

export const Disabled = Template.bind({});
Disabled.args = { disabled: true, text: 'Disabled Button' };

export const BareIcon = Template.bind({});
BareIcon.args = {
  variant: 'bare',
  text: 'Close',
  iconName: 'x',
  iconPosition: 'before',
  hideText: true
};

export const TextLinkButton = Template.bind({});
TextLinkButton.args = { variant: 'link', text: 'Text Link Button' };

export const TextLinkWithIcon = Template.bind({});
TextLinkWithIcon.args = {
  variant: 'link',
  text: 'Text Link Button',
  iconPosition: 'after',
  iconName: 'triangle-right'
};

export const Large = Template.bind({});
Large.args = { size: 'large', text: 'Large Button' };

export const Small = Template.bind({});
Small.args = { size: 'small', text: 'Small Button' };

export const Tiny = Template.bind({});
Tiny.args = { size: 'tiny', text: 'Tiny Button' };

export const Loading = Template.bind({});
Loading.args = { loading: true, text: 'Loading...' };

export const PrimaryLoading = Template.bind({});
PrimaryLoading.args = { variant: 'primary', loading: true, text: 'Loading...' };
