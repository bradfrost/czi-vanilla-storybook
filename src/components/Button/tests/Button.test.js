import React from 'react';
import { render } from 'enzyme';
import Button from '../Button';

it('renders a button', () => {
  const wrapper = render(<Button text="Click" />);
  expect(wrapper).toMatchSnapshot();
});

it('renders a link', () => {
  const wrapper = render(<Button text="Click" href="http://test.com" />);
  expect(wrapper).toMatchSnapshot();
});

it('handles variants', () => {
  const variants = [
    'primary',
    'secondary',
    'tertiary',
    'bare',
    'link',
    'link-inverted',
    'ghost'
  ];

  variants.forEach(variant => {
    const wrapper = render(<Button text="Click" variant={variant} />);
    expect(wrapper).toMatchSnapshot();
  });
});

it('handles sizes', () => {
  const sizes = ['large', 'small', 'tiny'];

  sizes.forEach(size => {
    const wrapper = render(<Button text="Click" size={size} />);
    expect(wrapper).toMatchSnapshot();
  });
});

it('handles types', () => {
  const types = ['button', 'reset', 'submit'];

  types.forEach(type => {
    const wrapper = render(<Button text="Click" type={type} />);
    expect(wrapper).toMatchSnapshot();
  });
});

it('can be inverted', () => {
  const wrapper = render(<Button text="Click" theme="inverted" />);
  expect(wrapper).toMatchSnapshot();
});

it('can be full width', () => {
  const wrapper = render(<Button text="Click" fullWidth />);
  expect(wrapper).toMatchSnapshot();
});

it('can be disabled', () => {
  const wrapper = render(<Button text="Click" disabled />);
  expect(wrapper).toMatchSnapshot();
});

it('can hide text', () => {
  const wrapper = render(<Button text="Click" hideText />);
  expect(wrapper).toMatchSnapshot();
});

it('can display loading', () => {
  const wrapper = render(<Button text="Click" loading />);
  expect(wrapper).toMatchSnapshot();
});

it('renders screen reader text', () => {
  const wrapper = render(<Button text="Click" screenReaderText="test" />);
  expect(wrapper).toMatchSnapshot();
});

it('renders icons', () => {
  const before = render(
    <Button text="Click" iconPosition="before" iconName="activity" />
  );
  expect(before).toMatchSnapshot();

  const after = render(
    <Button text="Click" iconPosition="after" iconName="activity" />
  );
  expect(after).toMatchSnapshot();
});
