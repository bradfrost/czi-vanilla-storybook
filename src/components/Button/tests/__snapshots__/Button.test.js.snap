// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be disabled 1`] = `
<button
  class="cn-c-btn"
  disabled=""
  tabindex="-1"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`can be full width 1`] = `
<button
  class="cn-c-btn cn-c-btn--block"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`can be inverted 1`] = `
<button
  class="cn-c-btn cn-c-btn--inverted"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`can display loading 1`] = `
<button
  class="cn-c-btn cn-is-loading"
>
  <svg
    aria-hidden="true"
    class="cn-c-icon cn-c-btn__icon"
    focusable="false"
  >
    <use
      href="test-file-stub#test-loader"
    />
  </svg>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`can hide text 1`] = `
<button
  class="cn-c-btn"
>
  <span
    class="cn-c-btn__text cn-u-is-vishidden"
  >
    Click
  </span>
</button>
`;

exports[`handles sizes 1`] = `
<button
  class="cn-c-btn cn-c-btn--large"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles sizes 2`] = `
<button
  class="cn-c-btn cn-c-btn--small"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles sizes 3`] = `
<button
  class="cn-c-btn cn-c-btn--tiny"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles types 1`] = `
<button
  class="cn-c-btn"
  type="button"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles types 2`] = `
<button
  class="cn-c-btn"
  type="reset"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles types 3`] = `
<button
  class="cn-c-btn"
  type="submit"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles variants 1`] = `
<button
  class="cn-c-btn cn-c-btn--primary"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles variants 2`] = `
<button
  class="cn-c-btn cn-c-btn--secondary"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles variants 3`] = `
<button
  class="cn-c-btn cn-c-btn--tertiary"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles variants 4`] = `
<button
  class="cn-c-btn cn-c-btn--bare"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles variants 5`] = `
<button
  class="cn-c-btn cn-c-btn--link"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles variants 6`] = `
<button
  class="cn-c-btn cn-c-btn--link-inverted"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`handles variants 7`] = `
<button
  class="cn-c-btn cn-c-btn--ghost"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`renders a button 1`] = `
<button
  class="cn-c-btn"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`renders a link 1`] = `
<a
  class="cn-c-btn"
  href="http://test.com"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</a>
`;

exports[`renders icons 1`] = `
<button
  class="cn-c-btn"
>
  <svg
    aria-hidden="true"
    class="cn-c-icon cn-c-btn__icon"
    focusable="false"
  >
    <use
      href="test-file-stub#activity"
    />
  </svg>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
</button>
`;

exports[`renders icons 2`] = `
<button
  class="cn-c-btn"
>
  <span
    class="cn-c-btn__text"
  >
    Click
  </span>
  <svg
    aria-hidden="true"
    class="cn-c-icon cn-c-btn__icon"
    focusable="false"
  >
    <use
      href="test-file-stub#activity"
    />
  </svg>
</button>
`;

exports[`renders screen reader text 1`] = `
<button
  class="cn-c-btn"
>
  <span
    class="cn-c-btn__text"
  >
    Click
    <span
      class="cn-u-is-vishidden"
    >
      test
    </span>
  </span>
</button>
`;
